File: 2025/day-10/part1.py
import itertools
import sys

if len(sys.argv) > 1:
	INPUT = sys.argv[1]
else:
	INPUT = "input1.txt"


def button_combination_patterns(
	coeffs: list[tuple[int, ...]],
) -> dict[tuple[int, ...], dict[tuple[int, ...], int]]:
	"""
	Group all possible button combinations by parity pattern (for indicators this IS the effect).

	For each parity pattern, store the minimal button press count needed to achieve it.
	"""
	num_indicators = len(coeffs[0])
	patterns_by_parity = {}

	# Try all 2^n subsets of buttons
	for num_buttons in range(len(coeffs) + 1):
		for button_indices in itertools.combinations(range(len(coeffs)), num_buttons):
			# Sum coefficients for selected buttons modulo 2 (XOR for binary)
			effect = tuple(
				sum(coeffs[i][j] for i in button_indices) % 2
				for j in range(num_indicators)
			)

			# Store minimal press count for this effect
			if effect not in patterns_by_parity:
				patterns_by_parity[effect] = num_buttons

	return patterns_by_parity


with open(INPUT) as f:
	machines_input = [line.strip().split(" ") for line in f.readlines()]
	machines = []
	for m in machines_input:
		indicator_seq = []
		for c in m[0][1:-1]:
			match c:
				case "#":
					indicator_seq.append(True)
				case ".":
					indicator_seq.append(False)
				case _:
					assert False
		seq_buttons = [
			tuple(int(j) for j in mj.strip("()").split(",")) for mj in m[1:-1]
		]
		seq_joltages = [int(j) for j in m[-1].strip("{}").split(",")]
		machines.append((tuple(indicator_seq), seq_buttons, seq_joltages))

	def process(
		indicator_seq: tuple[bool, ...],
		button_seq: list[tuple[int, ...]],
	) -> int:
		"""
		Find minimal button presses using precomputed combinations.
		For boolean indicators, each button combination produces a unique XOR pattern.
		"""
		# Convert indicator bools to integers (False=0, True=1)
		goal = tuple(int(b) for b in indicator_seq)

		# Convert buttons to binary coefficients (1 if button affects indicator, 0 otherwise)
		coeffs = [
			tuple(int(i in button) for i in range(len(indicator_seq)))
			for button in button_seq
		]

		patterns = button_combination_patterns(coeffs)

		# For boolean toggle problems, we just need to find the pattern that matches our goal
		if goal in patterns:
			return patterns[goal]
		else:
			return float("inf")

	print(
		sum([process(machines[i][0], machines[i][1]) for i, _ in enumerate(machines)])
	)

File: 2025/day-10/src/part1.rs
use std::collections::HashMap;

use nom::{
	IResult, Parser,
	character::complete::{self, line_ending, space1},
	multi::{fold_many1, separated_list1},
	sequence::delimited,
};

fn button_combination_patterns(
	coeffs: &[Vec<bool>],
	num_indicators: usize,
) -> HashMap<Vec<bool>, usize> {
	let mut patterns = HashMap::new();
	let num_buttons = coeffs.len();

	// Try all 2^n subsets of buttons (represented as bitmasks)
	for mask in 0u32..(1 << num_buttons) {
		// Build the XOR effect of all buttons enabled in `mask` using a fold
		let effect = (0..num_buttons)
			.filter(|&idx| (mask & (1u32 << idx)) != 0)
			.fold(vec![false; num_indicators], |mut acc, idx| {
				acc.iter_mut()
					.zip(&coeffs[idx])
					.for_each(|(e, &bit)| *e ^= bit);
				acc
			});

		// Store minimal press count for this effect pattern
		let num_presses = mask.count_ones() as usize;
		patterns
			.entry(effect)
			.and_modify(|existing: &mut usize| *existing = (*existing).min(num_presses))
			.or_insert(num_presses);
	}

	patterns
}

fn validate_button_sequences(machine: Machine) -> i32 {
	let num_indicators = machine.goal_indicator_seq.len();

	// Convert buttons to coefficient matrix (1 if button affects indicator, 0 otherwise)
	let coeffs: Vec<Vec<bool>> = machine
		.button_seq
		.iter()
		.map(|button| {
			(0..num_indicators)
				.map(|i| button.contains(&(i as u8)))
				.collect()
		})
		.collect();

	// Precompute all reachable patterns and their minimal press counts
	button_combination_patterns(&coeffs, num_indicators)
		.get(&machine.goal_indicator_seq)
		.copied()
		.unwrap_or(usize::MAX) as i32
}

#[tracing::instrument]
pub fn process(input: &str) -> miette::Result<String> {
	let (_, machines) = parse(input).expect("fine");
	let mut reaching_buttons = 0;
	for m in machines {
		reaching_buttons += validate_button_sequences(m);
	}
	Ok(reaching_buttons.to_string())
}

#[derive(Debug)]
struct Machine {
	goal_indicator_seq: Vec<bool>,
	button_seq: Vec<Vec<u8>>,
	joltage_seq: Vec<u16>,
}

fn parse(input: &str) -> IResult<&str, Vec<Machine>> {
	nom::multi::separated_list1(line_ending, |single_input| {
		let (single_input, goal_indicator_seq) = parse_goal_seq(single_input)?;
		// repeatedly skip spaces
		let (single_input, _) = complete::space1(single_input)?;
		// button parser with spaces
		let (single_input, button_seq) =
			separated_list1(space1, parse_button_seq).parse(single_input)?;
		let (single_input, _) = complete::space1(single_input)?;
		// joltage parser
		let (single_input, joltage_seq) = parse_joltage_seq(single_input)?;

		Ok((
			single_input,
			Machine {
				goal_indicator_seq,
				button_seq,
				joltage_seq,
			},
		))
	})
	.parse(input)
}
fn parse_goal_seq(input: &str) -> IResult<&str, Vec<bool>> {
	delimited(
		complete::char('['),
		fold_many1(
			nom::branch::alt((complete::char('.'), complete::char('#'))),
			Vec::new,
			|mut acc: Vec<bool>, state| {
				acc.push(match state {
					'.' => false,
					'#' => true,
					_ => panic!(),
				});
				acc
			},
		),
		complete::char(']'),
	)
	.parse(input)
}
fn parse_button_seq(input: &str) -> IResult<&str, Vec<u8>> {
	delimited(
		complete::char('('),
		separated_list1(complete::char(','), complete::u8),
		complete::char(')'),
	)
	.parse(input)
}

fn parse_joltage_seq(input: &str) -> IResult<&str, Vec<u16>> {
	delimited(
		complete::char('{'),
		separated_list1(complete::char(','), nom::character::complete::u16),
		complete::char('}'),
	)
	.parse(input)
}

#[cfg(test)]
mod tests {
	use super::*;

	#[test]
	fn test_process() -> miette::Result<()> {
		let input = "[.##.] (3) (1,3) (2) (2,3) (0,2) (0,1) {3,5,4,7}
[...#.] (0,2,3,4) (2,3) (0,4) (0,1,2) (1,2,3,4) {7,5,12,7,2}
[.###.#] (0,1,2,3,4) (0,3,4) (0,1,2,4,5) (1,2) {10,11,11,5,10,5}
";
		assert_eq!("7", process(input)?);
		Ok(())
	}
}
